src/passes/OnceReduction.cpp

```diff
@@ -0,0 +1,414 @@
+            // regardless of whether it is a global.set or a call.
+            //
+            // Note that assertions below verify that there are no children that
+            // we need to keep around, and so we can just nop the entire node.
+            ExpressionManipulator::nop(expr);
+          } else {
+            // From here on, this global is set, hopefully allowing us to
+            // optimize away others.
+            onceGlobalsWritten.insert(globalName);
+          }
+        };
+
+        if (auto* set = expr->dynCast<GlobalSet>()) {
+          if (optInfo.onceGlobals.at(set->name)) {
+            // This global is written.
+            assert(set->value->is<Const>());
+            optimizeOnce(set->name);
+          }
+        } else if (auto* call = expr->dynCast<Call>()) {
+          if (optInfo.onceFuncs.at(call->target).is()) {
+            // The global used by the "once" func is written.
+            assert(call->operands.empty());
+            optimizeOnce(optInfo.onceFuncs.at(call->target));
+            continue;
+          }
+
+          // This is not a call to a "once" func. However, we may have inferred
+          // that it definitely sets some "once" globals before it returns, and
+          // we can use that information.
+          for (auto globalName :
+               optInfo.onceGlobalsSetInFuncs.at(call->target)) {
+            onceGlobalsWritten.insert(globalName);
+          }
+        } else {
+          WASM_UNREACHABLE("invalid expr");
+        }
+      }
+    }
+
+    // As a result of the above optimization, we know which "once" globals are
+    // definitely written in this function. Regardless of whether this is a
+    // "once" function itself, that set of globals can be used in further
+    // optimizations, as any call to this function must set those.
+    // TODO: Aside from the entry block, we could intersect all the exit blocks.
+    optInfo.newOnceGlobalsSetInFuncs[func->name] =
+      std::move(onceGlobalsWrittenVec[0]);
+  }
+
+private:
+  OptInfo& optInfo;
+};
+
+} // anonymous namespace
+
+struct OnceReduction : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    OptInfo optInfo;
+
+    // Fill out the initial data.
+    for (auto& global : module->globals) {
+      // For a global to possibly be "once", it must be initialized to a
+      // constant. As we scan code we will turn this into false if we see
+      // anything that proves the global is not "once".
+      //   * Note that we don't check that the constant is zero - that is fine
+      //     for us to optimize, though it does indicate that the once function
+      //     will never ever run, which we could optimize further. TODO
+      //   * TODO: non-integer types?
+      optInfo.onceGlobals[global->name] = global->type.isInteger() &&
+                                          !global->imported() &&
+                                          global->init->is<Const>();
```

Comment:
Can't we further limit this to globals that are initialized to zero?

Reply:
Why does this have to be a constant? We don't care whether the value is zero or nonzero anyway.

Reply:
Hmm, good point! The initial value may as well be anything. Fixed.